Sources/ApodiniNetworking/HTTPServer.swift

```diff
@@ -0,0 +1,402 @@
+            case .shared:
+                break
+            case .createNew:
+                try! self.eventLoopGroup.syncShutdownGracefully()
+            }
+        }
+    }
+    
+    
+    /// Start the server.
+    /// This will attempt to open a NIO channel, bind it to the specified address, and set it up to handle incoming HTTP and HTTP2 requests, depending on the configuration options.
+    public func start() throws {
+        guard channel == nil else {
+            throw ApodiniNetworkingError(message: "Cannot start already-running servers")
+        }
+        let bootstrap = ServerBootstrap(group: eventLoopGroup)
+            .serverChannelOption(ChannelOptions.backlog, value: 256)
+            .serverChannelOption(ChannelOptions.socket(SocketOptionLevel(SOL_SOCKET), SO_REUSEADDR), value: 1)
+            .childChannelInitializer { [unowned self] (channel: Channel) -> EventLoopFuture<Void> in
+                logger.notice("Configuring NIO channel pipeline. TLS: \(tlsConfiguration != nil), HTTP/2: \(enableHTTP2)")
+                if let tlsConfig = tlsConfiguration {
+                    precondition(tlsConfig.applicationProtocols.contains("h2"), "h2 not found in \(tlsConfig.applicationProtocols)")
+                    let sslContext: NIOSSLContext
+                    do {
+                        sslContext = try NIOSSLContext(configuration: tlsConfig)
+                    } catch {
+                        logger.error("Unable to configure TLS: \(error)")
+                        return channel.close(mode: .all)
+                    }
+                    let tlsHandler = NIOSSLServerHandler(context: sslContext)
+                    return channel.pipeline.addHandler(tlsHandler)
+                        .flatMap { () -> EventLoopFuture<Void> in
+                            return channel.configureHTTP2SecureUpgrade { channel in
+                                channel.addApodiniNetworkingHTTP2Handlers(responder: self)
+                            } http1ChannelConfigurator: { channel in
+                                channel.addApodiniNetworkingHTTP1Handlers(responder: self)
+                            }
+                        }
+                        .flatMapError { error in
+                            channel.eventLoop.makeFailedFuture(error)
+                        }
+                } else {
+                    if enableHTTP2 {
+                        // NOTE this doesn't make sense and (probably) doesn't work
+                        return channel.addApodiniNetworkingHTTP2Handlers(responder: self)
+                    } else {
+                        return channel.addApodiniNetworkingHTTP1Handlers(responder: self)
+                    }
+                }
+            }
+            .childChannelOption(ChannelOptions.socket(IPPROTO_TCP, TCP_NODELAY), value: 1)
+            .childChannelOption(ChannelOptions.socket(SocketOptionLevel(SOL_SOCKET), SO_REUSEADDR), value: 1)
```

Comment:
First of all the `SO_REUSEADDR` seems to be duplicating the channel option above?
Secondly, why would be enable tcp no Delay and most importantly reuseaddr out of the box? The second one doesn't really make sense to me particularly. Might be beneficial in the future to provide some means of configuration of socket options?

Reply:
Yeah that's a leftover from a previous version of the code. I think that it might be a good idea to expose these through a config option, either at interface exporter level (i.e. that IEs can set this) or even at client level (i.e. that a WebServer could set this)